File: classic/Code/SearchInRotatedSortedArray.cpp
class Solution {
public:
    int search(int A[], int n, int target) {
        if(n==1){
            if(A[0]==target) {return 0;}
            else{return -1;}
        }
        int l=0;
        int r=n-1;
        while(l<=r){
            int m=(l+r)/2;
            if(m==l&&m==r){
                if(A[m]==target){return m;}
                else{return -1;}
            }
            if(A[m]==target){return m;}
            if(A[m]<target){
                if(A[m]<A[l]){
                    if(target<A[r]){
                        l=m+1;
                    }else if (target>A[r]){
                        r=m-1;
                    }else{
                        return r;
                    }
                }else{
                    l=m+1;
                }
            }else{
                if(A[m]>=A[l]){
                    if(target>A[l]){
                        r=m-1;
                    }else if (target<A[l]){
                        l=m+1;
                    }else{
                        return l;
                    }
                }else{
                    r=m-1;
                }
            }
        }
        return -1;
    }
};

File: classic/Code/SearchInRotatedSortedArray.java
public class Solution {
    int help(int [] A,int target,int start,int end){
        int mid=(start+end)/2;
        if(end-start<1){
            return -1;
        }
        if(end-start==1){
            if(A[start]==target){
                return start;
            }else{
                return -1;
            }
        }
        if(A[mid]==target){
            return mid;
        } 
        if(A[start]==target){
            return start;
        } 
        if(A[end-1]==target){
            return end-1;
        } 
        if (A[mid]>=A[start]){
            if(target>A[mid]){
                return help(A,target,mid+1,end-1);
            }else{
                return help(A,target,start+1,end-1);
            }
        }else {
            if(target<A[mid]){
                return help(A,target,start+1,mid);
            }else{
                return help(A,target,start+1,end-1);
            }
        }
    }
    public int search(int[] A, int target) {
        if(A.length==0){return -1;}
        return help(A,target,0,A.length);
    }
}
